x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HllFieldMapper.java

```diff
@@ -0,0 +1,737 @@
+            this.ignoreMalformed = gpfmMergeWith.ignoreMalformed;
+        }
+        if (precision.value() != gpfmMergeWith.precision.value()) {
+            conflicts.add("mapper [" + name() + "] has different [precision]");
+        } else if (precision.explicit() == false && gpfmMergeWith.precision.explicit()) {
+            this.precision = gpfmMergeWith.precision;
+            this.m = 1 << precision.value();
+        }
+    }
+
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) {
+        throw new UnsupportedOperationException("Parsing is implemented in parse(), this method should NEVER be called");
+    }
+
+    @Override
+    public ValueFetcher valueFetcher(MapperService mapperService, String format) {
+        if (format != null) {
+            throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] doesn't support formats.");
+        }
+        return new SourceValueFetcher(name(), mapperService, parsesArrayValue()) {
+            @Override
+            protected Object parseSourceValue(Object value) {
+                return value;
+            }
+        };
+    }
+
+    public static class HllFieldType extends MappedFieldType {
+
+        private final int precision;
+        private final boolean ignoreMalformed;
+
+        public HllFieldType(String name, boolean hasDocValues, int precision, boolean ignoreMalformed, Map<String, String> meta) {
+            super(name, false, hasDocValues, TextSearchInfo.SIMPLE_MATCH_ONLY, meta);
+            this.precision = precision;
+            this.ignoreMalformed = ignoreMalformed;
+        }
+
+        public int precision() {
+            return precision;
+        }
+
+        public boolean ignoreMalformed() {
+            return ignoreMalformed;
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName, Supplier<SearchLookup> searchLookup) {
+            failIfNoDocValues();
+            return (cache, breakerService, mapperService) ->
+                new IndexHllFieldData(name(), AnalyticsValuesSourceType.CARDINALITY) {
+
+                @Override
+                public LeafHllFieldData load(LeafReaderContext context) {
+
+                    return new LeafHllFieldData() {
+                        @Override
+                        public HllValues getHllValues() throws IOException {
+                            try {
+                                final BinaryDocValues values = DocValues.getBinary(context.reader(), fieldName);
+                                final ByteArrayDataInput dataInput = new ByteArrayDataInput();
+                                final InternalFixedLengthHllValue fixedValue = new InternalFixedLengthHllValue();
```

Comment:
Just for my own knowledge, is the idea here to allocate one of each encoding types, and then switch/reset between them as we encounter different encodings on each segment?

Reply:
Yes, that is the idea so when you read a value, read the first byte with the encoding type, reset the corresponding object and return it.